### PR TITLE
Add Demo for Certain Category

### LINKED ISSUE #5095: [Question] AutoComplete: How to create response similar to "Example of Certain Category"
I want to use AutoComplete in a manner similar to the "Example of Certain Category" image on https://ant.design/docs/spec/reaction#Lookup-Patterns: Customized input element and search results grouped in categories:

Unfortunately that is only an image, not a coded example. And the documentation has no info on how to create this style of display.

I have tried using `<AutoComplete.OptGroup>` which doesn't work. Also it seems combining custom input elements with options as children of `<AutoComplete>` does not work.

Is there any example of how to code this style of widget somewhere?

### PR DESCRIPTION
**Pull request by RaoHai**
+ close #5095
 + support OptGroup of AutoComplete

pre PR here, waitting for the design draft.

### CODE REVIEW

**Comment by JesperWe** [comment]
Thanks a lot, I think it was the missing docs for dataSource hierarchy with .children that I missed the most.

Some comments on the commit: 

- `optionLabelProp="text"` seems to be unused, and this prop is not in the API docs. 
- Using `dropdownMatchSelectWidth={false}` and `dropdownStyle={{ width: 300 }}` at the same time seems redundant, and these props are not in the API docs either.

**Comment by benjycui** [comment]
And we can add link to demo in https://ant.design/docs/spec/reaction#Lookup-Patterns

**Comment by RaoHai** [comment]
@JesperWe 

* `optionLabelProp` is the prop value of option which will be rendered as content of select.
* `dropdownMatchSelectWidth` means whether dropdown has same with with select field, when true, overriding with in `dropdownStyle` will not take effect.
* `AutoComplete` is implemented by `Select` recently. Some props are inherited.

I updated the document, now all the props of demo should have been used.

File: components/auto-complete/index.tsx [hunk 1]
```diff
@@ -94,7 +94,10 @@ export default class AutoComplete extends React.Component<AutoCompleteProps, any
 
     let options;
     const childArray = React.Children.toArray(children);
-    if (childArray.length && (childArray[0] as React.ReactElement<any>).type === Option) {
+    if (childArray.length && (
+        (childArray[0] as React.ReactElement<any>).type === Option ||
+        (childArray[0] as React.ReactElement<any>).type === OptGroup
```
Comment (yesmeck) [inline components/auto-complete/index.tsx]: https://github.com/ant-design/ant-design/issues/4655

File: components/auto-complete/demo/certain-category.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+  title: '文章',
+  children: [{
+    title: 'AntDesign 是一个设计语言',
+    count: 100000,
+  }],
+}];
+
+
+function renderTitle(title) {
+  return (<span>
```
Comment (yesmeck) [inline components/auto-complete/demo/certain-category.md]: 换行

**Review by yesmeck — CHANGES_REQUESTED** [review]

File: components/auto-complete/index.tsx (hunk 1 above)
Comment (RaoHai) [inline components/auto-complete/index.tsx]: 要改到 rc-select 里去？

Comment (yesmeck) [inline components/auto-complete/index.tsx]: 那这里应该用 `isSelectOption ` 和 `isSelectOptGroup` 判断吧。

File: components/auto-complete/index.tsx [hunk 3]
```diff
@@ -55,6 +55,16 @@ class InputElement extends React.Component<any, any> {
   }
 }
 
+interface OptionClass<p> extends React.ClassicComponentClass<p> {
+  isSelectOption: Boolean;
+  isSelectOptGroup: Boolean;
+}
+
+function isSelectOptionOrSelectOptGroup(child: React.ReactChild): Boolean {
+  const element = ((child as React.ReactElement<any>).type) as OptionClass<any>;
```
Comment (RaoHai) [inline components/auto-complete/index.tsx]: 好像只能这样写了。。

Comment (yesmeck) [inline components/auto-complete/index.tsx]: 直接 `(child as any).isSelectOption` 就可以了。

Comment (yesmeck) [inline components/auto-complete/index.tsx]: 直接上面的 `React.ReactChild` 换成 any 就可以了吧。

Comment (yesmeck) [inline components/auto-complete/index.tsx]: 但是要判断下`child.type`是不是存在。

Comment (yesmeck) [inline components/auto-complete/index.tsx]: 不需要判断。。。

Comment (RaoHai) [inline components/auto-complete/index.tsx]: 类型推导严谨一些嘛 😢

Comment (yesmeck) [inline components/auto-complete/index.tsx]: 判断 `child.type` 还是需要的，因为 `child` 可能是字符串之类的，`(child as React.ReactElement<any>)` 这里已经不严谨了其实，`child` 不一定是 `React.ReactElement` 的。

File: components/auto-complete/index.tsx [hunk 4]
```diff
@@ -55,6 +55,19 @@ class InputElement extends React.Component<any, any> {
...
+interface OptionClass<p> extends React.ClassicComponentClass<p> {
+  isSelectOption: Boolean;
+  isSelectOptGroup: Boolean;
+}
+
+function isSelectOptionOrSelectOptGroup(child: React.ReactChild): Boolean {
+  if (typeof child === 'string' || typeof child === 'number') {
+    return false;
+  }
+  const element = child.type as OptionClass<any>;
```
Comment (RaoHai) [inline components/auto-complete/index.tsx]: 这样吧， `type ReactChild = ReactElement<any> | string | number;`
排除掉 string 和 number 之后就可以安全的 .type 了

Comment (yesmeck) [inline components/auto-complete/index.tsx]: 感觉 `child` 的类型就是 `any`，`null`、`undefined`、`boolean` 都有可能的。

**Comment by RaoHai** [comment]
没啥问题的话我 rebase -i 一下然后合了吧

**Comment by yesmeck** [comment]
👌

**Pull request merged**